### PR TITLE
Issue #41 : travis CI configuration

### PR DESCRIPTION
**Pull request by WeinaJi**
Travis CI settings

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+addons:
+  # for Linux builds, we use APT
+  apt:
+    sources:
+      - ubuntu-toolchain-r-test
+    packages:
+      - flex
+      - bison
+      - cmake
+      - mpich
```
Comment (pramodk) [inline .travis.yml]: IIRC, this should be `libmpich-dev` (or `libopenmpi-dev`) ?

Comment (ohm314) [reply .travis.yml]: Yeap.. @WeinaJi please check what are exactly the package names on the respective system (you should be able to look this up on your workstation for example.

**Comment by WeinaJi** [comment]
This PR is finalized and ready for review. 
Could you please review it ? Thanks!

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+# Travis CoreNeuron settings
+#=============================================================================
+
+#=============================================================================
+# Environment
+#=============================================================================
+# Use new Travis infrastructure (Docker can't sudo yet)
+sudo: false
+language: cpp
+#PYTHON_VERSION=3.6.7
```
Comment (tristan0x) [inline .travis.yml]: Please remove dead code: #PYTHON_VERSION=3.6.7

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  # for Linux builds, we use APT
+  apt:
+    sources:
+      - ubuntu-toolchain-r-test
+    packages:
+      - flex
+      - bison
+      - libopenmpi-dev
+      - openmpi-bin
+#      - cmake  #cmake 3.12.4 on travis
```
Comment (tristan0x) [inline .travis.yml]: Same here, please dead code

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+#      - python3-dev
+#      - python3-pip
+
+  # for Mac builds, we use Homebrew
+  homebrew:
+    packages:
+      - flex
+      - bison
+      - openmpi
+#      - cmake  #cmake 3.11.3 on travis
```
Comment (tristan0x) [inline .travis.yml]: same here.
Also please sort list of packages

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+#=============================================================================
+# Common Packages
+#=============================================================================
+addons:
+  # for Linux builds, we use APT
+  apt:
+    sources:
+      - ubuntu-toolchain-r-test
+    packages:
+      - flex
```
Comment (tristan0x) [inline .travis.yml]: please sort list of packages

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+#=============================================================================
+# Install dependencies : flex
+#=============================================================================
+before_install:
+  # brew installed flex and bison is not in $PATH
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
+    export PATH=/usr/local/opt/flex/bin:/usr/local/opt/bison/bin:$PATH;
+    else
+    pyenv global $PYTHON_VERSION;
```
Comment (tristan0x) [inline .travis.yml]: I don't think the `pyenv ...` and `ls /usr/bin` is required. If so can you remove it?

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+#=============================================================================
+script:
+  - echo "------- Build, Test and Install -------"
+  - mkdir build && cd build
+  - cmake ${cmake_option} -DCMAKE_INSTALL_PREFIX=$HOME/CoreNeuron ..
+  - cmake --build .
+  - ctest --output-on-failure
+  - make install
+
+#=============================================================================
```
Comment (tristan0x) [inline .travis.yml]: Please remove the notifications section if unused

Comment (WeinaJi) [reply .travis.yml]: Thanks!. I have implemented your comments, removed deadcode and unneeded lines.

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+# Travis CoreNeuron settings
+#=============================================================================
+
+#=============================================================================
+# Environment
+#=============================================================================
+# Use new Travis infrastructure (Docker can't sudo yet)
```
Comment (ohm314) [inline .travis.yml]: Does this comment make sense, given that `suda: false` ?

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+#=============================================================================
+# Build matrix
+#=============================================================================
+matrix:
+  fast_finish: true
+  include:
+    - os: linux
+      dist: xenial
+      env:
+        - cmake_option="-DCORENRN_ENABLE_MPI=ON"
```
Comment (ohm314) [inline .travis.yml]: Could you please now also add the builds with the NMODL backend?

Comment (ohm314) [reply .travis.yml]: Additionally, currently you only build with gcc. Could you please also build with clang? For both, could you please choose recent versions? gcc 9.2 and clang 9.0

File: .travis.yml (hunk 8 above)
Comment (WeinaJi) [reply .travis.yml]: It seems that "sudo" key is deprecated in travis so I would remove it and that comment.

File: .travis.yml (hunk 9 above)
Comment (WeinaJi) [reply .travis.yml]: I have rebased my branch on the "cmake" branch to include the sub module NMODL. 
The latest version of the travis CI config contains two more build tests:
1. CORENRN_ENABLE_NMODL=ON 
2. build with linux gcc 9
The source tool chain for Clang 9 is currently not available in Travis. An older version should work, I can add it if needed.

**Comment by pramodk** [comment]
very nice! only my query is about [boost not being found and unit tests are being skipped](https://travis-ci.org/BlueBrain/CoreNeuron/jobs/604369217#L441):

```
CMake Warning at /usr/local/cmake-3.12.4/share/cmake-3.12/Modules/FindBoost.cmake:577 (message):
433  Imported targets and dependency information not available for Boost version
434  (all versions older than 1.33)
435Call Stack (most recent call first):
436  /usr/local/cmake-3.12.4/share/cmake-3.12/Modules/FindBoost.cmake:963 (_Boost_COMPONENT_DEPENDENCIES)
437  /usr/local/cmake-3.12.4/share/cmake-3.12/Modules/FindBoost.cmake:1622 (_Boost_MISSING_DEPENDENCIES)
438  tests/CMakeLists.txt:16 (find_package)
439
440
441-- Boost not found, unit tests disabled
```

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  # for Linux builds, we use APT
+  apt:
+    sources:
+      - ubuntu-toolchain-r-test
+    packages:
+      - bison
+      - flex
+      - libopenmpi-dev
+      - openmpi-bin
+      - libboost-all-dev
```
Comment (pramodk) [inline .travis.yml]: @WeinaJi : do we need to duplicate `addons` section here? I see that some of those being defined at line 61.

**Comment by pramodk** [comment]
Please retest

**Comment by WeinaJi** [comment]
Hello all, 

I have added the build with ISPC backend. Please review it. Thanks!

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  - if [[ "$USE_NMODL" == "ON" || "$USE_ISPC" == "ON" ]]; then
+        pip3 install jinja2 pyyaml pytest sympy;
+    fi
+  - if [ -n "$GCC_VERSION" ]; then
+        export CXX="g++-${GCC_VERSION}" CC="gcc-${GCC_VERSION}";
+        $CXX --version;
+    fi
+  # install ISPC compiler
+  - if [[ "$USE_ISPC" == "ON" ]]; then
+        wget https://sourceforge.net/projects/ispcmirror/files/v1.12.0/ispc-v1.12.0-linux.tar.gz;
```
Comment (pramodk) [inline .travis.yml]: Just last one minor thing: could you take care of `platform` here? i.e. if someone enabled ISPC on OSX then link should be for macOS.

Comment (WeinaJi) [reply .travis.yml]: I have added the source link for mac os, shall I also add an ISPC test for mac ? Currently it only tests with linux.

Comment (pramodk) [reply .travis.yml]: nitpick : can we avoid duplicate code for wget/untar by something like:

```bash
if [ "$TRAVIS_OS_NAME" == "linux" ]; then url_os="linux"; else url_os="macOS"; fi
url="https://sourceforge.net/projects/ispcmirror/files/v1.12.0/ispc-v1.12.0-${url_os}.tar.gz"
wget $url;
```

Comment (pramodk) [reply .travis.yml]: its not necessary to have ISPC test for mac.

**Pull request merged**